relayer/strategies.go

```diff
@@ -1,58 +1,241 @@
-				ap, err := UnrelayedAcknowledgements(src, dst)
-				if err != nil {
-					src.Log(fmt.Sprintf("unrelayed acks error: %s", err))
-				} else {
-					if len(ap.Src) > 0 && src.debug {
-						src.Log(fmt.Sprintf("[%s] unrelayed-acks-> %v", src.ChainID(), ap.Src))
-					}
-					if len(ap.Dst) > 0 && dst.debug {
-						dst.Log(fmt.Sprintf("[%s] unrelayed-acks-> %v", dst.ChainID(), ap.Dst))
-					}
-					if !ap.Empty() {
-						if err = RelayAcknowledgements(src, dst, ap, maxTxSize, maxMsgLength); err != nil && src.debug {
-							src.Log(fmt.Sprintf("relay acks error: %s", err))
-						}
+				// Filter for open channels that are not already in our slice of open channels
+				// TODO implement a filter list of channels we want to relay against or a list of channels to ignore
+				srcOpenChannels = FilterOpenChannels(srcChannels, srcOpenChannels)
+
+				// Spin up a goroutine to relay packets & acks for each channel that isn't already being relayed against
+				for _, channel := range srcOpenChannels {
+					if !channel.active {
+						channel.active = true
+						go RelayUnrelayedPacketsAndAcks(src, dst, maxTxSize, maxMsgLength, channel, channels, doneChan)
 					}
 				}
 
-				time.Sleep(100 * time.Millisecond)
+				for channel := range channels {
+					channel.active = false
+					break
+				}
+
+				// Make sure we are removing channels no longer in OPEN state from the slice of open channels
+				for i, channel := range srcOpenChannels {
+					if channel.channel.State != types.OPEN {
+						srcOpenChannels[i] = srcOpenChannels[len(srcOpenChannels)-1]
+						srcOpenChannels = srcOpenChannels[:len(srcOpenChannels)-1]
+					}
+				}
 			}
 		}
 	}()
 	return func() { doneChan <- struct{}{} }, nil
 }
+
+// QueryChannelsOnConnection queries all the channels associated with a connection on the src chain.
+func QueryChannelsOnConnection(src *Chain) ([]*types.IdentifiedChannel, error) {
+	// Query the latest heights on src & dst
+	srch, err := src.ChainProvider.QueryLatestHeight()
+	if err != nil {
+		return nil, err
+	}
+
+	// Query the list of channels for the connection on src
+	var srcChannels []*types.IdentifiedChannel
+
+	if err = retry.Do(func() error {
+		srcChannels, err = src.ChainProvider.QueryConnectionChannels(srch, src.ConnectionID())
+		return err
+	}, RtyAtt, RtyDel, RtyErr, retry.OnRetry(func(n uint, err error) {
+		src.LogRetryQueryConnectionChannels(n, err, src.ConnectionID())
+	})); err != nil {
+		return nil, err
+	}
+
+	return srcChannels, nil
+}
+
+// FilterOpenChannels takes a slice of channels and adds all the channels with OPEN state to a new slice of channels.
+// NOTE: channels will not be added to the slice of open channels more than once.
+func FilterOpenChannels(channels []*types.IdentifiedChannel, openChannels []*ActiveChannel) []*ActiveChannel {
+	inSlice := false
+
+	// Filter for open channels
+	for _, channel := range channels {
+		if channel.State == types.OPEN {
```

Comment:
I think you can move the declaration of `inSlice` here, so that its lifecycle is limited to one iteraction of `for _, channel := range channels` and we don't have to explicitly re-assign it at the end of the loop.